src/main/scala/org/apache/spark/sql/delta/storage/COSLogStore.scala

```diff
@@ -0,0 +1,61 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.delta.storage
+
+import java.io.IOException
+import java.nio.charset.StandardCharsets.UTF_8
+import java.nio.file.FileAlreadyExistsException
+
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.{FileSystem, Path}
+import org.apache.spark.SparkConf
+
+/**
+ * LogStore implementation for IBM Cloud Object Storage.
+ *
+ * We assume the following from COS's [[FileSystem]] implementations:
+ * - Write on COS is all-or-nothing, whether overwrite or not.
+ * - Write is atomic.
```

Comment:
LogStore also requires `Mutual exclusion: Only one writer must be able to create (or rename) a file at the final destination.`. What's the behavior when multiple writes are happening at the same time?

In addition, according to the doc `enable or disable atomic writes to COS using Etags. When the flag fs.cos.fast.upload is set true atomic writes will be available only for writes that require only one block.`, looks like atomic write doesn't work for large files?

Reply:
The implementation of atomic write use Etags for concurrency control.
The high level flow is:
1. Take HEAD of object at path to check if it exists and if yes get Etag `e`.
- If object doesn't exist then do a PUT using header `If-None-Match: *`.
- Will only succeed if this is the first PUT to that object name.
- If this fails because of constraint (concurrent write) then throw an illegal state . 
2. If object exists and `overwrite == false` throw file already exists exception.
3. If object exists and `overwrite == true`, PUT object using header `If-Match: e`.
- If this fails because of constraint (concurrent write) then throw an illegal state exception.

So in case of multiple writers at the same time only one will succeed while the others will get an exception.
For example see this exception:
```
Exception in thread "Thread-14" java.lang.IllegalStateException: Failed due to concurrent write
        at org.apache.spark.sql.delta.storage.COSLogStore.write(COSLogStore.scala:53)
        at org.apache.spark.sql.delta.storage.LogStoreTests$$anon$1.run(LogStoreTests.scala:43)
Caused by: java.io.IOException: saving output tmp/delta_try1/_delta_log/0 com.amazonaws.services.s3.model.AmazonS3Exception: At least one of the preconditions you specified did not hold. (Service: Amazon S3; Status Code: 412; Error Code: PreconditionFailed; Request ID: 5c8dd3e6-f6b9-4a17-9d13-88ff7a274d04; S3 Extended Request ID: null), S3 Extended Request ID: null
        at com.ibm.stocator.fs.cos.COSOutputStream.close(COSOutputStream.java:197)
        at org.apache.hadoop.fs.FSDataOutputStream$PositionCache.close(FSDataOutputStream.java:72)
        at org.apache.hadoop.fs.FSDataOutputStream.close(FSDataOutputStream.java:106)
        at org.apache.spark.sql.delta.storage.COSLogStore.write(COSLogStore.scala:50)
        ... 1 more
```
<br>
As for large files - The use of Etags is supported currently for `PutObjectRequest` therefore it means that the atomic write support is available only when the write is done in one chunk.
The documentation wasn't clear about that so I have updated it accordingly (the documentation in the connector will be updated as well):

```
Note: Atomic write is available only when the write is done in one chunk. 
- When the flag `fs.cos.fast.upload` is set to `false` atomic write will be available only for write that is lower or equal to `fs.cos.multipart.threshold size` (by default max integer). 
- When the flag `fs.cos.fast.upload` is set to `true` atomic write will be available only for write that require only one block .
```

So we expect writes that go through the LogStore to be written in one chunk by the connector in order to achieve mutual exclusion.
Note that using the default configuration (`fs.cos.fast.upload=false`) write is done in one chunk for files of about ~2.15GB so it should be reasonable.

Reply:
Adding to the previous comment, looking at the current code these are the places where a write is done using the LogStore API:

1. Writing the `_last_checkpoint` file - https://github.com/delta-io/delta/blob/7c2c0dabff5fb48b2e08c2b1c9ef1ffbedf11c95/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala#L122

2. Writing the actions during commit -
- https://github.com/delta-io/delta/blob/7c2c0dabff5fb48b2e08c2b1c9ef1ffbedf11c95/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala#L401
- https://github.com/delta-io/delta/blob/7c2c0dabff5fb48b2e08c2b1c9ef1ffbedf11c95/src/main/scala/org/apache/spark/sql/delta/commands/ConvertToDeltaCommand.scala#L361

3. Writing manifest file for Presto/Athena - https://github.com/delta-io/delta/blob/7c2c0dabff5fb48b2e08c2b1c9ef1ffbedf11c95/src/main/scala/org/apache/spark/sql/delta/hooks/GenerateSymlinkManifest.scala#L243

1 is a small fixed sized object, and for 2 and 3 I think it's reasonable to assume that these writes should be below the default threshold of ~2.15GB to ensure these writes are done in one chunk and so can be atomic. If needed in special cases the user can configure the threshold differently.

In addition regarding the write of the checkpoint: 
https://github.com/delta-io/delta/blob/7c2c0dabff5fb48b2e08c2b1c9ef1ffbedf11c95/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala#L228
since write in COS is not partially visible this should be OK as well.

In addition I have added checks to the LogStore implementation to make sure the write will be atomic.

Reply:
IBM COS supports now conditional headers on multipart uploads as well so this is no longer necessary.
The PR and proposed doc was updated to remove these checks.